### DISCUSSION TITLE
Does setting up an index automatically mean that it is used for record methods?

### DISCUSSION THREAD

**Discussion by usrme** [pocketbase/pocketbase#2544] [Q&A] [answered]
First, thank you for the wonderful job you've done on creating and maintaining this. PocketBase is fantastic!

Now, if I've set up `CREATE UNIQUE INDEX asdf ON hosts (hostname)` on the collection `hosts` (through the UI, of course) and I use `app.Dao().FindRecordByData("hosts", "hostname", "somehostname")` elsewhere **will it automatically use the index**?

The reason I'm asking is because I kind of implemented a little cache look-up using `app.Cache().Get()` and `app.Cache().Set()`, but later on when I remembered the indexing functionality I realized the caching logic may have been a mistake.

**Comment by ganigeorgiev** [comment, marked as answer]
> will it automatically use the index?

`FindFirstRecordByData` will generate a query like:

```sql
SELECT hosts.* 
FROM hosts
WHERE hostname = 'somehostname'
LIMIT 1
```

So I expect SQLite to use the index, but keep in mind that for more complex queries this may not be the case as the query planner may decide to optimize the execution differently (_eg. in case there is an `ORDER BY` statement or other where condition_).

**Reply by usrme** [reply]
Thank you! I'll do some testing to see if there's any noticeable difference between one and the other, though I do suspect the caching to be overkill. As an approximation: would _you_ expect hitting `app.Cache().Get()` to be faster than going through SQLite?

**Reply by ganigeorgiev** [reply]
> would you expect hitting app.Cache().Get() to be faster than going through SQLite?

I expect `app.Cache().Get()` to be always faster as it doesn't do much work - internally it is a direct map lookup with a rlock, so I guess the `interface` type assertion would be the only "slow" part when using it.

But in general caching comes with higher maintenance cost and my recommendation would be to avoid it until you hit some limitation and there is no other easier alternative.

_Note that official testbed with various collections configurations and benchmarks will be available soon._

**Reply by usrme** [reply]
Thanks! The caching solution was indeed faster by about 10x on my own machine with just a single data point for both executions for 50 records, but in terms of absolute numbers (0.00181s vs. 0.0188s) there wasn't enough of a difference for me to keep my custom caching logic.
